### PR TITLE
Merge from master to Dev_Unity_2017.2.0

### PR DESCRIPTION
**Pull request by keveleigh**
I'd recommend looking through this commit by commit. My merge commits (and the master commits they contain) should be obvious.

### CODE REVIEW

**Comment by StephenHodgson** [comment]
I am very nervous about this merge haha

**Comment by keveleigh** [comment]
@StephenHodgson As am I haha...we've waited so long to do this, I'm not sure a better way to go about it.

**Comment by StephenHodgson** [comment]
I'd like to do smaller chunks if possible.  Like do one feature area at a time.

**Comment by keveleigh** [comment]
@StephenHodgson The problem with that is that I'm bringing in all the history. Can't really pick and choose feature areas on a merge, at least without ruining history for any future merges...

**Comment by StephenHodgson** [comment]
Did you do a rebase of the 2017.2 branch onto the master branch?

**Comment by keveleigh** [comment]
@StephenHodgson I did not. Hrm, that's probably a better way to go about this...I only did straight merges.

**Comment by NeerajW** [comment]
I'll hold off until this:
https://help.github.com/articles/about-git-rebase

**Comment by keveleigh** [comment]
The rebase (and my original merges) is a little more difficult due to the inclusion of commit history without content in #714 (the commits are coming up as removal of the files/changes when merged back into the Dev2017.2 branch).

In progress though!

**Comment by StephenHodgson** [comment]
Sorry, that's my fault you're having a difficult time with that.

**Comment by StephenHodgson** [comment]
Maybe I'm setting up my scene wrong, but I keep getting issues with the pointer assignment on start with this PR.

**Comment by keveleigh** [comment]
@StephenHodgson Ahh, that's a result of my migration of 92384c50a693a192fb5a6de9f0331ce930c8d541. Since OnEnable() gets called before Start(), Pointer is null when this gets called on app startup. Adding a null check, then the errors above disappear.

**Comment by StephenHodgson** [comment]
I've got one last revert to make from `StageManager.cs`

**Comment by StephenHodgson** [comment]
Forgot I accidentally removed that bool `updateStageBounds`.

**Comment by jessemcculloch** [comment]
What is our plan for testing this?

**Comment by StephenHodgson** [comment]
I tried building today but have ran into multiple issues.

I wasn't able to run this from the editor with the Acer Headset & controllers, but it ran fine without them.

Didn't get a chance to test builds again the HoloLens yet.

**Comment by keveleigh** [comment]
@jessemcculloch At this point, it might be best to wait until our merge into master of this branch. That'll make the diff much simpler, as we'll only be looking at the changes made in this branch, instead of re-reviewing all new additions to master, like this PR is showing.

**Comment by jessemcculloch** [comment]
Either way, does not matter to me.  I am using the 2017.2 branch for my development as is right now, since I want the Immersive Headset stuff, so I don't mind testing, I just don't want to duplicate work.  This is such a large job, It's only going to get worse trying to test it the longer it hangs out there and other stuff gets pushed to master.

Let me know if there is anything you guys need me to do.

**Comment by StephenHodgson** [comment]
I too don't mind testing. We're really pushing to have the headsets working with our projects here too.

**Comment by NeerajW** [comment]
If we are not merging master into this, should this PR be closed?

**Comment by StephenHodgson** [comment]
I'm still interested in a piece meal approach and doing a PR for each version.
We can also create patches for recent PRs just on this branch, and just start over with master.  The initial changes were from build last year and don't really have history attached to them, correct?

**Comment by DDReaper** [comment]
I would also probably recommend closing and redoing this PR once the other breaking change merges have been done on master

**Comment by DDReaper** [comment]
I agree with @NeerajW this should be closed now in favour of your progressive approach @StephenHodgson

**Comment by StephenHodgson** [comment]
@keveleigh I'd just go ahead and merge this.

File: .gitignore [hunk 1]
```diff
@@ -38,4 +39,6 @@ obj/
 *.idb
 *.opendb
 WSATestCertificate.pfx
-WSATestCertificate.pfx.meta
\ No newline at end of file
+WSATestCertificate.pfx.meta
+
+!*SRMesh.obj
```
Comment (StephenHodgson) [inline .gitignore]: What's with the custom obj ref?

Should we allow obj extensions in general? (Pretty much the same as .fbx)

Comment (keveleigh) [reply .gitignore]: .obj is also used for some Visual Studio generated files, but I think we might catch all those with the obj/ folder ignore. It's probably safe to remove all *.obj ignores, negating the need for this specific one

Comment (StephenHodgson) [reply .gitignore]: Agreed. I've never had issues with VS obj objects in my projects with the Obj/ folder ignore.

**Comment by StephenHodgson** [comment]
@keveleigh, let's make a tag for 2017.2 dev and then merge this.

**Comment by keveleigh** [comment]
@StephenHodgson Make a tag pre-merge?

**Comment by StephenHodgson** [comment]
Yeah, I think that'll help out anyone who's not ready to update yet.

**Comment by StephenHodgson** [comment]
@keveleigh I made the tag already.

**Comment by keveleigh** [comment]
@StephenHodgson I think we should wait for #1007 to go in, then merge this.

**Comment by StephenHodgson** [comment]
I disagree. We need to move this.

**Comment by keveleigh** [comment]
@StephenHodgson I'll merge the latest master changes in here, then merge. We'll likely break the other PR though.

**Comment by StephenHodgson** [comment]
We just need to take the steps to get there.  This branch is already really far behind, and the U2017.2 release is quickly approaching, and a lot of us have demos we need to complete before Oct 3rd.

**Comment by StephenHodgson** [comment]
On the bright side, the Headsets haven't really made it out to many Devs yet, but the longer we wait the harder it'll be to make changes without breaking other people's projects.

**Comment by DDReaper** [comment]
Here, here.
I'd rather this gets up to date sooner and then stabilise before we are too long into Oct, before Mr is fully released.
Need to be able to have something devs can pick up and run with asap

**Comment by StephenHodgson** [comment]
@keveleigh, I don't mind doing the latest master changes, because it involves the Xbox input.  I'd like to work on that if possible.

**Comment by keveleigh** [comment]
Did you squash this PR? The point of this was to bring in all the actual commits from master, but now source control still thinks there are merges to be done. I'm going to redo this without squashing.

**Comment by keveleigh** [comment]
Did you squash this PR? The point of a merge is to bring in all the actual commits, but now source control still thinks there are merges to be done. I'm going to redo this without squashing.

**Comment by StephenHodgson** [comment]
@keveleigh, sorry it must have been set by default from the last time I merged. (I guess it remembers your last action?)

**Comment by keveleigh** [comment]
@StephenHodgson All good! It'll be easy to re-do

**Comment by keveleigh** [comment]
Looks like my data may have just taken a bit to catch up. I see the commit history now, hmm. Not sure what happened before.

**Pull request merged**